### ISSUE TITLE
Unable to recover from Kafka connection error by using Supervision.Directive

### ISSUE THREAD

**Issue by xafos** [akka/alpakka-kafka#650]
I have the following code where if the connection with Kafka fails, the consumer dies without using the Supervision.Directive: strategyConsumed returns false.

```  // return the source
  def getSource()(implicit actorSystem: ActorSystem): Source[(CommittableOffset,ConsumerRecord[Any,Any]) , Consumer.Control] = {
    Consumer.committableSource(buildConsumerSettings(consumerConfig), Subscriptions.topics(topicName))
      .map(commitableMessage => {
        val consumerRecord = deserializeMessage(TopicName("topicName"), commitableMessage.record)
        (commitableMessage.committableOffset, consumerRecord)
      })
  }

  // define the strategy
  var strategyConsumed = false
  val strategyDecision = new function.Function[Throwable, Directive] {
    override def apply(t: Throwable): Directive = {
      strategyConsumed = true
      t match {
        case _ => Supervision.resume
      }
    }
  }

  val promise = Promise[Int]

  materializer = ActorMaterializer(createActorMaterializerSettings(strategyDecider, inputBuffer, system))

  // get the source
  source = new KafkaSourceFactory(consumerConfig).getSource()

  source
    .map(msg => {
      val commitableOffset = msg._1
      val eventualCommit = commitableOffset.commitScaladsl()
      Await.result(eventualCommit, 5 seconds)
      promise.success( msg._2.value().toString.toInt)
    })
    .to(Sink.foreach(value => {
      println(value)
    }))
    .run()(materializer)

  control.shutdown()

  // awaiting response
  Await.result(promise.future, 30 seconds)

  // fails with timeout and strategyConsumed = false
  assert(strategyConsumed)```

**Comment by ennru** [comment]
Please have a look at [this example](https://doc.akka.io/docs/akka-stream-kafka/current/serialization.html#spray-json) in the documentation that shows a working solution for what you are trying to do.

**Comment by ennru** [comment]
Errors from connectivity problems have been improved in the latest 1.0-M1 release and fail the stream with the cause now.

**Comment by politrons** [comment]
@ennru  Hi  enru, thanks for the response and link. 
We have coverage the problem of the de/serialization time ago using byteArray and then do it the encode/decode in a map to control the Error with the StrategyDecision.

Here the problem we`re facing is not about deserialization, is about that when our Consumer try to connect to Kafka, Kafka is temporally unavailable, so we start receiving wakeupExceptions, until we reach the max number of retries and our Consumer actor die, but the StrategyDecision in any case is invoked.
The only workaround so far has been to put the max retries of the wakeup as 0, which means it will try forever.

But the idea was that has a hook/callback as the StrategyDecision to notice if after the max retry number I cannot connect do another action.

**Comment by ennru** [comment]
I'm not sure it is good news, but we'll upgrade to the Kafka 2.x client soon and won't use `WakeupExceptions` anymore. 
The possibly bad part is, that we won't be able to notice if there is a Kafka broker to talk to - in your case that's what you want, I guess.

**Comment by politrons** [comment]
Yes, so no I can asume with your answer then that there’s no mechanism to make the supervisor strategy when the consumer cannot talk with Kafka broker. I’m wondering if you get rid of the wakeupexception it does not means that you get rid of the max number of wake ups. And if you does, I hope you guys provide another mechanism to continue trying to connect forever as you do right now. In any case if that change affects my code I will have to see great new feature to make me want to go to this new major version. And I hope you guys keep maintaining the previous 1.x with bug fixing for a while.

**Comment by ennru** [comment]
All wakeup-handling will go away with #614.
If we don't do anything special, the consumer will just continue to try to talk to the broker forever. 
We might build something to detect connectivity as I just suggested in #674.

**Comment by politrons** [comment]
Well I hope they will listen to you. But I think remove the max wake up time was not necessary a great move. In some scenarios maybe some people really wants their consumer dies to do something else. The right move IMHO it would be put the retry number to forever by default.

I will subscribe to your proposal to keep me posted.
Regards
